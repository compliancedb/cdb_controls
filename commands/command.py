import json


class Command:

    def __init__(self, context):
        self._context = context

    def execute(self):
        print("MERKELY_COMMAND={}".format(self.command))
        self.concrete_execute()

    @property
    def command(self):
        return self._env("MERKELY_COMMAND")

    @property
    def api_token(self):
        return self._env("MERKELY_API_TOKEN")

    @property
    def host(self):
        return self._env("MERKELY_HOST")

    @property
    def merkelypipe(self):
        MERKELYPIPE_PATH = "/Merkelypipe.json"
        with open(MERKELYPIPE_PATH) as file:
            return json.load(file)

    def _env(self, name):
        return self._context['env'].get(name, None)
