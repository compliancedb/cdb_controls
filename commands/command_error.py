
class CommandError(Exception):
    def __init__(self, status_code, message):
        super().__init__(message)
        self._status_code = status_code

    @property
    def status_code(self):
        return self._status_code
