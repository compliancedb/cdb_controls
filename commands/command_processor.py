from cdb.api_schema import ApiSchema
from cdb.http import http_put_payload
import json


def execute(context):
    command = context['env'].get("MERKELY_COMMAND", None)
    if command == "declare_pipeline":
        DeclarePipelineCommand(context).execute()
    if command == "log_artifact":
        LogArtifactCommand(context).execute()
    return 0


class Command:

    def __init__(self, context):
        self._context = context

    def execute(self):
        print("MERKELY_COMMAND={}".format(self.command))
        self.concrete_execute()

    @property
    def command(self):
        return self._env("MERKELY_COMMAND")

    @property
    def api_token(self):
        return self._env("MERKELY_API_TOKEN")

    @property
    def host(self):
        return self._env("MERKELY_HOST")

    @property
    def merkelypipe(self):
        MERKELYPIPE_PATH = "/Merkelypipe.json"
        with open(MERKELYPIPE_PATH) as file:
            return json.load(file)

    def _env(self, name):
        return self._context['env'].get(name, None)


class DeclarePipelineCommand(Command):

    def concrete_execute(self):
        pipelines_url = ApiSchema.url_for_pipelines(self.host, self.merkelypipe)
        http_put_payload(url=pipelines_url, payload=self.merkelypipe, api_token=self.api_token)


class LogArtifactCommand(Command):

    def concrete_execute(self):
        fingerprint = self._env("MERKELY_FINGERPRINT")
        context = self._context
        context['description'] = "Created by build " + self._env('MERKELY_CI_BUILD_NUMBER')
        context['git_commit'] = self._env('MERKELY_ARTIFACT_GIT_COMMIT')
        context['commit_url'] = self._env('MERKELY_ARTIFACT_GIT_URL')
        context['build_url'] = self._env('MERKELY_CI_BUILD_URL')
        context['is_compliant'] = self._env('MERKELY_IS_COMPLIANT') == "TRUE"

        FILE_PROTOCOL = "file://"
        if fingerprint.startswith(FILE_PROTOCOL):
            self._log_artifact_file(fingerprint[len(FILE_PROTOCOL):])

        DOCKER_PROTOCOL = "docker://"
        if fingerprint.startswith(DOCKER_PROTOCOL):
            self._log_artifact_docker_image(fingerprint[len(DOCKER_PROTOCOL):])

    def _log_artifact_file(self, filename):
        context = self._context
        pathed_filename = '/' + filename
        print("Getting SHA for file:// artifact: " + pathed_filename)
        artifact_sha = context['sha_digest_for_file'](pathed_filename)
        print("Calculated digest: " + artifact_sha)
        # print("Publish artifact to ComplianceDB")
        print('MERKELY_IS_COMPLIANT: ' + str(context['is_compliant']))
        create_artifact(self.api_token, self.host, self.merkelypipe,
                        artifact_sha, pathed_filename,
                        context['description'],
                        context['git_commit'],
                        context['commit_url'],
                        context['build_url'],
                        context['is_compliant'])

    def _log_artifact_docker_image(self, image_name):
        context = self._context
        print("Getting SHA for docker:// artifact: " + image_name)
        artifact_sha = context['sha_digest_for_docker_image'](image_name)
        print("Calculated digest: " + artifact_sha)
        # print("Publish artifact to ComplianceDB")
        print('MERKELY_IS_COMPLIANT: ' + str(context['is_compliant']))
        create_artifact(self.api_token, self.host, self.merkelypipe,
                        artifact_sha, image_name,
                        context['description'],
                        context['git_commit'],
                        context['commit_url'],
                        context['build_url'],
                        context['is_compliant'])



def create_artifact(api_token, host, merkelypipe,
                    sha256, filename,
                    description, git_commit, commit_url, build_url, is_compliant):
    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = ApiSchema.url_for_artifacts(host, merkelypipe)
    http_put_payload(url, create_artifact_payload, api_token)
