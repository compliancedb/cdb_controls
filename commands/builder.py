from commands import *

COMMANDS = {
    "declare_pipeline": DeclarePipeline,
    "log_artifact": LogArtifact,
    "log_deployment": LogDeployment,
    "log_evidence": LogEvidence,
    "log_approval": LogApproval,
    "log_test": LogTest,
    "control_deployment": ControlDeployment,
}


def build_command(name):
    klass = COMMANDS.get(name)
    if klass is None:
        raise CommandError(f"Unknown command: {name}")
    return klass
