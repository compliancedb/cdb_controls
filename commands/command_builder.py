from commands import *


def build_command(context):
    name = Command(context).name.value
    if name == "declare_pipeline":
        return DeclarePipelineCommand(context)
    if name == "log_artifact":
        return LogArtifactCommand(context)
    if name == "log_deployment":
        return LogDeploymentCommand(context)
    if name == "log_evidence":
        return LogEvidenceCommand(context)
    return None
