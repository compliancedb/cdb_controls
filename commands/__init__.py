from .command_error import CommandError

from .json_loader import load_json

from .context import Context

from .command import Command

from .declare_pipeline import DeclarePipeline
from .control_deployment import ControlDeployment
from .log_artifact import LogArtifact
from .log_deployment import LogDeployment
from .log_evidence import LogEvidence
from .log_approval import LogApproval
from .log_test import LogTest

from .command_builder import build_command, COMMANDS
from .runner import run, main
