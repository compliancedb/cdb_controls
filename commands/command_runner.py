from commands import *


def run(context):
    try:
        cmd = build_command(context)
        if cmd is not None:
            print(f"MERKELY_COMMAND={cmd.name.value}")
            for env_var in list(cmd.env_vars):
                env_var.verify()
            return cmd()
    except CommandError as exc:
        print(f"Error: {str(exc)}")
        return 144


