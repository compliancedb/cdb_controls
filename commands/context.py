import os
from commands import CommandError
from fingerprinters import *


class Context:
    """
    Holds (some) external dependencies to make Classic style testing easier.
    """
    def __init__(self, env=None,
                 docker_fingerprinter=None,
                 file_fingerprinter=None,
                 sha256_fingerprinter=None):
        self.__env = env
        self.__docker_fingerprinter = docker_fingerprinter
        self.__file_fingerprinter = file_fingerprinter
        self.__sha256_fingerprinter = sha256_fingerprinter

        if self.__env is None:
            self.__env = os.environ

        if self.__docker_fingerprinter is None:
            self.__docker_fingerprinter = DockerFingerprinter()

        if self.__file_fingerprinter is None:
            self.__file_fingerprinter = FileFingerprinter()

        if self.__sha256_fingerprinter is None:
            self.__sha256_fingerprinter = Sha256Fingerprinter()

    @property
    def env(self):
        return self.__env

    def fingerprinter_for(self, string):
        d = self.__docker_fingerprinter
        f = self.__file_fingerprinter
        s = self.__sha256_fingerprinter
        if d.handles_protocol(string):
            return d
        elif f.handles_protocol(string):
            return f
        elif s.handles_protocol(string):
            return s
        else:
            raise CommandError(f"Unknown protocol: {string}")

