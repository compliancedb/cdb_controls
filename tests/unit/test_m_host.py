from commands import command_runner, Context
from tests.utils import *


def test_host_not_set_defaults(capsys):
    ev = core_env_vars()
    ev.pop("MERKELY_HOST")

    with dry_run(ev) as env, scoped_merkelypipe_json():
        command_runner.run(Context(env))

    verify_approval(capsys)


def test_host_set(capsys):
    ev = core_env_vars()
    ev["MERKELY_HOST"] = "https://test.merkely.com"

    with dry_run(ev) as env, scoped_merkelypipe_json():
        command_runner.run(Context(env))

    verify_approval(capsys)
