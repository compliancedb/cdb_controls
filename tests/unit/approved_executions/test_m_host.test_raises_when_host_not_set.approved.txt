MERKELY_COMMAND=declare_pipeline
ERROR: MERKELY_HOST environment-variable not set
